File: checkpoint.go
package main

type CheckpointResult struct {
	Messages        []BeeperMessage
	NextCheckpoint  string
	CheckpointFound bool
	Bootstrapped    bool
}

func relevantIncoming(m BeeperMessage) bool {
	return !m.IsSender && !m.IsDeleted && m.Type != "REACTION"
}

// MessagesAfterCheckpoint accepts Beeper's newest-first page and returns unseen incoming messages oldest-first.
func MessagesAfterCheckpoint(items []BeeperMessage, checkpoint string) CheckpointResult {
	newestIncoming := ""
	for _, m := range items {
		if relevantIncoming(m) {
			newestIncoming = m.ID
			break
		}
	}
	if checkpoint == "" {
		return CheckpointResult{NextCheckpoint: newestIncoming, Bootstrapped: true}
	}
	index := -1
	for i, m := range items {
		if m.ID == checkpoint {
			index = i
			break
		}
	}
	if index < 0 {
		if newestIncoming == "" {
			newestIncoming = checkpoint
		}
		return CheckpointResult{NextCheckpoint: newestIncoming}
	}
	var messages []BeeperMessage
	for i := index - 1; i >= 0; i-- {
		if relevantIncoming(items[i]) {
			messages = append(messages, items[i])
		}
	}
	if newestIncoming == "" {
		newestIncoming = checkpoint
	}
	return CheckpointResult{Messages: messages, NextCheckpoint: newestIncoming, CheckpointFound: true}
}

File: src/checkpoint.ts
import type { BeeperMessage } from "./types.js";

export type CheckpointResult = {
  messages: BeeperMessage[];
  nextCheckpoint?: string;
  checkpointFound: boolean;
  bootstrapped: boolean;
};

function isRelevantIncoming(message: BeeperMessage): boolean {
  return !message.isSender && !message.isDeleted && message.type !== "REACTION";
}

/** Accepts Beeper's newest-first page and returns unseen incoming messages oldest-first. */
export function messagesAfterCheckpoint(
  newestFirst: BeeperMessage[],
  checkpoint?: string,
): CheckpointResult {
  const incoming = newestFirst.filter(isRelevantIncoming);
  const newestIncoming = incoming[0]?.id;

  if (!checkpoint) {
    return {
      messages: [],
      nextCheckpoint: newestIncoming,
      checkpointFound: false,
      bootstrapped: true,
    };
  }

  const checkpointIndex = newestFirst.findIndex((message) => message.id === checkpoint);
  if (checkpointIndex === -1) {
    // A stale checkpoint must never cause a reply storm over an entire page of history.
    return {
      messages: [],
      nextCheckpoint: newestIncoming ?? checkpoint,
      checkpointFound: false,
      bootstrapped: false,
    };
  }

  const messages = newestFirst.slice(0, checkpointIndex).filter(isRelevantIncoming).reverse();
  return {
    messages,
    nextCheckpoint: newestIncoming ?? checkpoint,
    checkpointFound: true,
    bootstrapped: false,
  };
}
